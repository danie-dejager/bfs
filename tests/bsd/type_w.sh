# Only ffs supports whiteouts on FreeBSD
command -v mdconfig &>/dev/null || skip
command -v newfs &>/dev/null || skip

cleanup=()
do_cleanup() {
    # Run cleanup hooks in reverse order
    while ((${#cleanup[@]} > 0)); do
        cmd="${cleanup[-1]}"
        unset 'cleanup[-1]'
        eval "bfs_sudo $cmd"
    done
}
trap do_cleanup EXIT

clean_scratch

# Create a ramdisk
truncate -s1M scratch/img
md=$(bfs_sudo mdconfig scratch/img) || skip
cleanup+=("mdconfig -du $md")

# Make an ffs filesystem
bfs_sudo newfs -n "/dev/$md" >&2 || skip
mkdir scratch/mnt

# Mount it
bfs_sudo mount "/dev/$md" scratch/mnt || skip
cleanup+=("umount scratch/mnt")

# Make it owned by us
bfs_sudo chown "$(id -u):$(id -g)" scratch/mnt
"$XTOUCH" -p scratch/mnt/{lower/{foo,bar,baz},upper/{bar,baz/qux}}

# Mount a union filesystem within it
bfs_sudo mount -t unionfs -o below scratch/mnt/{lower,upper}
cleanup+=("umount scratch/mnt/upper")

# Create a whiteout
rm scratch/mnt/upper/bar

# FreeBSD find doesn't have -printf, so munge -ls output
munge_ls() {
    sed -En 's|.*([-drwx]{10}).*(scratch/.*)|'"$1"': \1 \2|p'
}

# Do a few tests in one
{
    # Normally, we shouldn't see the whiteouts
    invoke_bfs scratch/mnt -ls | munge_ls 1
    # -type w adds whiteouts to the output
    invoke_bfs scratch/mnt -type w -ls | munge_ls 2
    # So this is not the same as test 1
    invoke_bfs scratch/mnt \( -type w -or -not -type w \) -ls | munge_ls 3
    # Unmount the unionfs
    bfs_sudo umount scratch/mnt/upper
    unset 'cleanup[-1]'
    # Now repeat the same tests
    invoke_bfs scratch/mnt -ls | munge_ls 4
    invoke_bfs scratch/mnt -type w -ls | munge_ls 5
    invoke_bfs scratch/mnt \( -type w -or -not -type w \) -ls | munge_ls 6
} >"$OUT"
sort_output
diff_output
