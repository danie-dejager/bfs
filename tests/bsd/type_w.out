1: -rw-r--r-- scratch/mnt/lower/bar
1: -rw-r--r-- scratch/mnt/lower/baz
1: -rw-r--r-- scratch/mnt/lower/foo
1: -rw-r--r-- scratch/mnt/upper/baz/qux
1: -rw-r--r-- scratch/mnt/upper/foo
1: drwxr-xr-x scratch/mnt
1: drwxr-xr-x scratch/mnt/lower
1: drwxr-xr-x scratch/mnt/upper
1: drwxr-xr-x scratch/mnt/upper/baz
2: w--------- scratch/mnt/upper/bar
3: -rw-r--r-- scratch/mnt/lower/bar
3: -rw-r--r-- scratch/mnt/lower/baz
3: -rw-r--r-- scratch/mnt/lower/foo
3: -rw-r--r-- scratch/mnt/upper/baz/qux
3: -rw-r--r-- scratch/mnt/upper/foo
3: drwxr-xr-x scratch/mnt
3: drwxr-xr-x scratch/mnt/lower
3: drwxr-xr-x scratch/mnt/upper
3: drwxr-xr-x scratch/mnt/upper/baz
3: w--------- scratch/mnt/upper/bar
4: -rw-r--r-- scratch/mnt/lower/bar
4: -rw-r--r-- scratch/mnt/lower/baz
4: -rw-r--r-- scratch/mnt/lower/foo
4: -rw-r--r-- scratch/mnt/upper/baz/qux
4: drwxr-xr-x scratch/mnt
4: drwxr-xr-x scratch/mnt/lower
4: drwxr-xr-x scratch/mnt/upper
4: drwxr-xr-x scratch/mnt/upper/baz
5: w--------- scratch/mnt/upper/bar
6: -rw-r--r-- scratch/mnt/lower/bar
6: -rw-r--r-- scratch/mnt/lower/baz
6: -rw-r--r-- scratch/mnt/lower/foo
6: -rw-r--r-- scratch/mnt/upper/baz/qux
6: drwxr-xr-x scratch/mnt
6: drwxr-xr-x scratch/mnt/lower
6: drwxr-xr-x scratch/mnt/upper
6: drwxr-xr-x scratch/mnt/upper/baz
6: w--------- scratch/mnt/upper/bar
